power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +284,597 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
-        // TODO(mgovers): implement outter loop tap changer
-        throw PowerGridError{};
+    auto optimize(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                  CalculationMethod method) const -> ResultType {
+        initialize(regulator_order);
+
+        if (auto result = iterate(state, regulator_order, method); strategy_ == OptimizerStrategy::any) {
+            return result;
+        }
+
+        // refine solution
+        step_all_back(regulator_order);
+        return iterate(state, regulator_order, method);
+    }
+
+    auto iterate(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                 CalculationMethod method) const -> ResultType {
+        auto result = calculate_with_retry(state, method);
+
+        bool tap_changed = true;
+        while (tap_changed) {
+            tap_changed = false;
+            UpdateBuffer update_data;
+
+            for (auto const& same_rank_regulators : regulator_order) {
+                for (auto const& regulator : same_rank_regulators) {
+                    tap_changed = tap_changed || control_transformer(regulator, state, result, update_data);
+                }
+                if (tap_changed) {
+                    break;
+                }
+            }
+            if (tap_changed) {
+                update_state(update_data);
+                result = calculate_(state, method);
+            }
+        }
+
+        add_tap_regulator_output(state, regulator_order, result);
+
+        return result;
+    }
+
+    auto calculate_with_retry(State const& state, CalculationMethod method) const {
+        try {
+            return calculate_(state, method);
+        } catch (SparseMatrixError const&) {
+            return calculate_(state, CalculationMethod::linear);
+        } catch (IterationDiverge const&) {
+            return calculate_(state, CalculationMethod::linear);
+        }
+    }
+
+    bool control_transformer(RegulatedTransformer const& regulator, State const& state, ResultType const& math_output,
+                             UpdateBuffer& update_data) const {
+        bool tap_changed = false;
+
+        regulator.transformer.apply([&](transformer_c auto const& transformer) {
+            using TransformerType = std::remove_cvref_t<decltype(transformer)>;
+            using sym = typename ResultType::value_type::sym;
+
+            auto const param = regulator.regulator.get().template calc_param<sym>();
+            auto const node_state =
+                NodeState<sym>{.u = u_pu_controlled_node<TransformerType>(regulator, state, math_output),
+                               .i = i_pu_controlled_node<TransformerType>(regulator, state, math_output)};
+
+            auto const cmp = node_state <=> param;
+            auto new_tap_pos = [&transformer, &cmp] {
+                if (cmp > 0) { // NOLINT(modernize-use-nullptr)
```

Comment:
will be fixed in clang-tidy 18.1.0 https://releases.llvm.org/18.1.0/tools/clang/tools/extra/docs/clang-tidy/checks/modernize/use-nullptr.html#cmdoption-arg-IgnoredTypes

hopefully it will be included in Ubuntu 24.04 quite soon, but it's unlikely to be included in 22.04